## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [28]:
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
columns_to_drop = ['EIN', 'NAME']
application_df = application_df.drop(columns=columns_to_drop)


In [3]:
# Determine the number of unique values in each column.
unique_value_counts = application_df.nunique()
print(unique_value_counts)


APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [4]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_type_counts)

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64


In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 3
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = application_type_counts[application_type_counts <= cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
T13         66
T12         27
T2          16
Other       11
Name: APPLICATION_TYPE, dtype: int64

In [6]:
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [7]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()

print(classification_counts)

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64


In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_counts_greater_than_1 = classification_counts[classification_counts > 1]
print(classification_counts_greater_than_1)


C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64


In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = classification_counts[classification_counts <= cutoff_value].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
Other       45
C1230       36
C1400       34
C2300       32
C7200       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1278       10
C1238       10
C1235        9
C1237        9
C7210        7
C4100        6
C1720        6
C2400        6
C1257        5
C1600        5
Name: CLASSIFICATION, dtype: int64

In [10]:
# Convert categorical data to numeric with `pd.get_dummies`
categorical_columns = ['APPLICATION_TYPE','CLASSIFICATION','AFFILIATION','USE_CASE','ORGANIZATION','STATUS','INCOME_AMT','SPECIAL_CONSIDERATIONS','ASK_AMT']

application_df = pd.get_dummies(application_df, columns=categorical_columns)


In [17]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop(columns=['IS_SUCCESSFUL'])  
y = application_df['IS_SUCCESSFUL']  

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [18]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [14]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = X_train_scaled.shape[1]

nn = Sequential()

# First hidden layer
nn.add(Dense(units=32, input_dim=input_features, activation='relu'))

# Second hidden layer
nn.add(Dense(units=16, activation='relu'))

# Output layer
nn.add(Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 32)                282464    
                                                                 
 dense_1 (Dense)             (None, 16)                528       
                                                                 
 dense_2 (Dense)             (None, 1)                 17        
                                                                 
Total params: 283009 (1.08 MB)
Trainable params: 283009 (1.08 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


In [21]:
from tensorflow.keras.callbacks import ModelCheckpoint


In [26]:
checkpoint_filepath = 'model_checkpoint.h5'
model_checkpoint_callback = ModelCheckpoint(
    filepath=checkpoint_filepath,
    save_weights_only=True,
    save_freq=5
)

In [27]:
# Train the model
history = nn.fit(X_train_scaled, y_train, epochs=50, verbose=1,callbacks=[model_checkpoint_callback])


Epoch 1/50
858/858 [==============================] - 24s 28ms/step - loss: 0.3840 - accuracy: 0.8172
Epoch 2/50
858/858 [==============================] - 25s 29ms/step - loss: 0.3838 - accuracy: 0.8173
Epoch 3/50
858/858 [==============================] - 24s 28ms/step - loss: 0.3838 - accuracy: 0.8174
Epoch 4/50
858/858 [==============================] - 25s 30ms/step - loss: 0.3833 - accuracy: 0.8173
Epoch 5/50
858/858 [==============================] - 25s 29ms/step - loss: 0.3831 - accuracy: 0.8181
Epoch 6/50
858/858 [==============================] - 22s 26ms/step - loss: 0.3831 - accuracy: 0.8184
Epoch 7/50
858/858 [==============================] - 23s 26ms/step - loss: 0.3831 - accuracy: 0.8176
Epoch 8/50
858/858 [==============================] - 22s 26ms/step - loss: 0.3829 - accuracy: 0.8183
Epoch 9/50
858/858 [==============================] - 22s 26ms/step - loss: 0.3831 - accuracy: 0.8177
Epoch 10/50
858/858 [==============================] - 23s 27ms/step - loss: 0.382

In [28]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - loss: 0.7387 - accuracy: 0.7143 - 332ms/epoch - 2ms/step
Loss: 0.7386752963066101, Accuracy: 0.7142857313156128


In [29]:
# Export our model to HDF5 file
model_filename = 'AlphabetSoupCharity.h5'
nn.save(model_filename)


d:\RDA\envs\dev00\lib\site-packages\keras\src\engine\training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


## Optimized Model

In [30]:
from tensorflow.keras.layers import Dense, Dropout, LeakyReLU
from tensorflow.keras.optimizers import Adam

In [31]:
# Define the model
optimized_model = Sequential()

In [32]:
# Adding hidden layers with more neurons and different activation functions
optimized_model.add(Dense(units=64, input_dim=input_features))
optimized_model.add(LeakyReLU(alpha=0.2))
optimized_model.add(Dense(units=32))
optimized_model.add(LeakyReLU(alpha=0.2))

In [33]:
# Adding a dropout layer to reduce overfitting
optimized_model.add(Dropout(0.3))

# Output layer
optimized_model.add(Dense(units=1, activation='sigmoid'))

# Compile the model with a different optimizer and learning rate
optimizer = Adam(learning_rate=0.001)
optimized_model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])


In [34]:
history = optimized_model.fit(X_train_scaled, y_train, epochs=30, verbose=1)


Epoch 1/30
858/858 [==============================] - 8s 8ms/step - loss: 0.6794 - accuracy: 0.6847
Epoch 2/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4396 - accuracy: 0.7982
Epoch 3/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4109 - accuracy: 0.8091
Epoch 4/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4063 - accuracy: 0.8102
Epoch 5/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4031 - accuracy: 0.8095
Epoch 6/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4021 - accuracy: 0.8107
Epoch 7/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4014 - accuracy: 0.8108
Epoch 8/30
858/858 [==============================] - 7s 8ms/step - loss: 0.4003 - accuracy: 0.8112
Epoch 9/30
858/858 [==============================] - 7s 8ms/step - loss: 0.3997 - accuracy: 0.8101
Epoch 10/30
858/858 [==============================] - 8s 9ms/step - loss: 0.3994 - accuracy: 0.8102

In [35]:
# Evaluate the model using the test data
model_loss, model_accuracy = optimized_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

215/215 - 0s - loss: 0.6229 - accuracy: 0.7262 - 492ms/epoch - 2ms/step
Loss: 0.6229126453399658, Accuracy: 0.7262390851974487


In [37]:
# Export our model to HDF5 file
model_filename = 'AlphabetSoupCharity_Optimization.h5'
optimized_model.save(model_filename)

d:\RDA\envs\dev00\lib\site-packages\keras\src\engine\training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


## Conclusion Summaries

*first run, no checkpoint at 50 epochs* \
215/215 - 0s - loss: 0.6556 - accuracy: 0.7165 - 487ms/epoch - 2ms/step \
Loss: 0.6555503010749817, Accuracy: 0.716472327709198 

*second run, checkpoints at 50 epochs* \
215/215 - 0s - loss: 0.7387 - accuracy: 0.7143 - 332ms/epoch - 2ms/step \
Loss: 0.7386752963066101, Accuracy: 0.7142857313156128 

*third run, optimized no checkpoint, at 30 epochs* \
215/215 - 0s - loss: 0.6229 - accuracy: 0.7262 - 492ms/epoch - 2ms/step \
Loss: 0.6229126453399658, Accuracy: 0.7262390851974487
